File: ueueo-framework/ueueo-threading/src/main/java/Abp/Threading/AsyncLocalAmbientDataContext.cs
using System.Collections.Concurrent;
using System.Threading;
using Volo.Abp.DependencyInjection;

namespace Volo.Abp.Threading;

public class AsyncLocalAmbientDataContext : IAmbientDataContext, ISingletonDependency
{
    private static readonly ConcurrentDictionary<String, AsyncLocal<Object>> AsyncLocalDictionary = new ConcurrentDictionary<String, AsyncLocal<Object>>();

    public void SetData(String key, Object value)
    {
        var asyncLocal = AsyncLocalDictionary.GetOrAdd(key, (k) => new AsyncLocal<Object>());
        asyncLocal.Value = value;
    }

    public Object GetData(String key)
    {
        var asyncLocal = AsyncLocalDictionary.GetOrAdd(key, (k) => new AsyncLocal<Object>());
        return asyncLocal.Value;
    }
}

File: abp-frameworks/ueueo-threading/src/main/java/com/ueueo/threading/AsyncLocalAmbientDataContext.java
package com.ueueo.threading;

import java.util.concurrent.ConcurrentHashMap;

public class AsyncLocalAmbientDataContext implements IAmbientDataContext {

    private static final ConcurrentHashMap<String, ThreadLocal<Object>> ASYNC_LOCAL_DICTIONARY = new ConcurrentHashMap<>();

    @Override
    public void setData(String key, Object value) {
        ThreadLocal<Object> asyncLocal = ASYNC_LOCAL_DICTIONARY.computeIfAbsent(key, s -> new ThreadLocal<>());
        asyncLocal.set(value);
    }

    @Override
    public Object getData(String key) {
        ThreadLocal<Object> asyncLocal = ASYNC_LOCAL_DICTIONARY.computeIfAbsent(key, s -> new ThreadLocal<>());
        return asyncLocal.get();
    }
}
